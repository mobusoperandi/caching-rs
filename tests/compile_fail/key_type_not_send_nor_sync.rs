use michie::memoized;
use std::rc::Rc;
use std::collections::HashMap;

fn generic_in_impl() {
    #[memoized(key_expr = input, store_type = HashMap<Rc<()>, ()>)]
    fn f(input: Rc<()>) -> () {
        ()
    }
}

fn main() {}
