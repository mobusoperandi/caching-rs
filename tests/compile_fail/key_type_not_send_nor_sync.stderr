error[E0277]: `Rc<()>` cannot be shared between threads safely
 --> tests/compile_fail/key_type_not_send_nor_sync.rs:6:5
  |
6 |     #[memoized(key_expr = input, store_type = HashMap<Rc<()>, ()>)]
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Rc<()>` cannot be shared between threads safely
  |
  = help: within `(Rc<()>, ())`, the trait `Sync` is not implemented for `Rc<()>`
  = note: required because it appears within the type `(Rc<()>, ())`
  = note: required because of the requirements on the impl of `Sync` for `hashbrown::raw::RawTable<(Rc<()>, ())>`
  = note: required because it appears within the type `hashbrown::map::HashMap<Rc<()>, (), RandomState>`
  = note: required because it appears within the type `HashMap<Rc<()>, ()>`
  = note: required for the cast from `HashMap<Rc<()>, ()>` to the object type `dyn Any + Send + Sync`
  = note: this error originates in the attribute macro `memoized` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `Rc<()>` cannot be sent between threads safely
 --> tests/compile_fail/key_type_not_send_nor_sync.rs:6:5
  |
6 |     #[memoized(key_expr = input, store_type = HashMap<Rc<()>, ()>)]
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Rc<()>` cannot be sent between threads safely
  |
  = help: within `(Rc<()>, ())`, the trait `Send` is not implemented for `Rc<()>`
  = note: required because it appears within the type `(Rc<()>, ())`
  = note: required because of the requirements on the impl of `Send` for `hashbrown::raw::RawTable<(Rc<()>, ())>`
  = note: required because it appears within the type `hashbrown::map::HashMap<Rc<()>, (), RandomState>`
  = note: required because it appears within the type `HashMap<Rc<()>, ()>`
  = note: required for the cast from `HashMap<Rc<()>, ()>` to the object type `dyn Any + Send + Sync`
  = note: this error originates in the attribute macro `memoized` (in Nightly builds, run with -Z macro-backtrace for more info)
